Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatable.cs

```diff
@@ -0,0 +1,407 @@
+                    case State.MovingRotatingScaling:
+                    case State.Scaling:
+                    case State.Rotating:
+                    case State.MovingScaling:
+                        OnTwoHandManipulationStarted(newState);
+                        break;
+                }
+                switch (oldState)
+                {
+                    case State.Start:
+                        OnManipulationStarted();
+                        break;
+                    case State.Scaling:
+                    case State.Rotating:
+                    case State.RotatingScaling:
+                    case State.MovingRotatingScaling:
+                    case State.MovingScaling:
+                        OnTwoHandManipulationEnded();
+                        break;
+                }
+            }
+            else
+            {
+                switch (newState)
+                {
+                    case State.Moving:
+                        OnOneHandMoveUpdated();
+                        break;
+                    case State.Scaling:
+                    case State.Rotating:
+                    case State.RotatingScaling:
+                    case State.MovingRotatingScaling:
+                    case State.MovingScaling:
+                        OnTwoHandManipulationUpdated();
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+
+        private void OnTwoHandManipulationUpdated()
+        {
+            var targetRotation = HostTransform.rotation;
+            var targetPosition = HostTransform.position;
+            var targetScale = HostTransform.localScale;
+
+            if ((currentState & State.Moving) > 0)
+            {
+                targetPosition = m_moveLogic.Update(GetHandsCentroid(), targetPosition);
+            }
+            if ((currentState & State.Rotating) > 0)
+            {
+                targetRotation = m_rotateLogic.Update(m_handsPressedLocationsMap, HostTransform, targetRotation);
+            }
+            if ((currentState & State.Scaling) > 0)
+            {
+                targetScale = m_scaleLogic.Update(m_handsPressedLocationsMap);
+            }
+
+            HostTransform.position = targetPosition;
+            HostTransform.rotation = targetRotation;
+            HostTransform.localScale = targetScale;
+        }
+
+        private void OnOneHandMoveUpdated()
+        {
+            var targetPosition = m_moveLogic.Update(m_handsPressedLocationsMap.Values.First(), HostTransform.position);
+
+            HostTransform.position = targetPosition;
+        }
+
+        private void OnTwoHandManipulationEnded()
+        {
+        }
+
+        private Vector3 GetHandsCentroid()
+        {
+            Vector3 result = m_handsPressedLocationsMap.Values.Aggregate(Vector3.zero, (current, state) => current + state);
+            return result / m_handsPressedLocationsMap.Count;
```

Comment:
Probably not because we'd have to have hands detected to ever get them into the map.

But it's still probably good to assert the count != 0.

If there's no way it can be 0 we should just use 
```
return m_handsPressedLocationsMap.Values.Aggregate(Vector3.zero, (current, state) => current + state);
```

Reply:
Although imo we should stay away from linq and just iterate over the map.